python/smqtk/algorithms/saliency/rise.py

```diff
@@ -0,0 +1,117 @@
+                                    anti_aliasing=False)[x:x + input_size[0], y:y + input_size[1]]
+        self.masks = masks.reshape(-1, *input_size, 1)
+    
+    def augment(self, image_mat):
+        """
+        :param numpy.ndarray image_mat:
+            Image matrix to be augmented.
+
+        :return: A numpy arrays of augmented image matrices as well as masks
+            that indicate the regions in the augmented images that are
+            unmodified with respect to the input image (preserved regions).
+
+            Returned augmented images should be in the dimension format
+            [index, height, width [,channel]] with the the same data type as
+            the input image matrix.
+
+            Returned masks should be in the dimension format
+            [index, height, width] with the boolean data type.
+        :rtype: (numpy.ndarray, numpy.ndarray)
+        """
+        # If image is grayscale
+        if len(image_mat.shape) == 2:
+            image_mat = np.expand_dims(image_mat, 2).repeat(3, axis=2)
+        return self.masks * image_mat, self.masks
+            
+        
+class RISEGenerator (ImageSaliencyMapGenerator):
+    def generate(self, base_image, augmenter, descriptor_generator,
+                 blackbox):
+        """
+        Generate an image saliency heat-map matrix given a blackbox's behavior
+        over the descriptions of an augmented base image.
+
+        :param numpy.ndarray image_mat:
+            Numpy image matrix of the format [height, width [,channel]] that is
+            to be augmented.
+
+        :param ImageSaliencyAugmenter augmenter:
+            Augmentation algorithm following
+            the :py:class:`ImageSaliencyAugmenter` interface.
+
+        :param smqtk.algorithms.DescriptorGenerator descriptor_generator:
+            A descriptor generation algorithm following
+            the :py:class:`smqtk.algorithms.DescriptorGenerator` interface.
+
+        :param SaliencyBlackbox blackbox:
+            Blackbox algorithm implementation following
+            the :py:class:`SaliencyBlackbox` interface.
+
+        :return: A :py:class:`numpy.ndarray` matrix of the same [height, width]
+            shape as the input image matrix but of floating-point type within
+            the range of [0,1], where areas of higher value represent more
+            salient regions according to the given blackbox algorithm.
+        :rtype: numpy.ndarray[float]
+        """
+        # augmenter = RISEAugmenter(1000, 10, 0.5, (224, 224))
+        masked_images, masks = augmenter.augment(base_image)
+        
+        scores = np.zeros(masked_images.shape[0])
+        # Performance-critical part since we run the model on a large number of images.
+        # Should be well-optimized, and should be done in batches if possible.
+        for i, masked_image in enumerate(masked_images):
+            desc = descriptor_generator.compute_descriptor(masked_image)
```

Comment:
masked_image cannot be an input to compute_descriptor() as this is not consistent with the API, convert it to smqtk.representation.DataElement format or a iterator of DataMemoryElement before using compute_descriptor() or compute_descriptor_async(). Please refer to the class "Logit_ImageSaliencyMapGenerator" on saliency-implementations branch for a closer look.